### DISCUSSION TITLE
How to fix query's underlying capacity based on some input?

### DISCUSSION THREAD

**Discussion by frederikhors** [ent/ent#2645] [Q&A]
With Bun I was using code like this:

```go
output := make([]Entity, 0, len(ids))

err := db.NewSelect().Model(&output).Query()
```

What to do now that I'm using Ent?

Is there a way to tell Ent to create the query's underlying slice with capacity based on `len(ids)`?

If I use `Limit()` I'm changing the query too... I don't know if is 100% correct.

**Comment by a8m** [comment]
Hey! What are you trying to achieve, reduce allocations?

**Reply by frederikhors** [reply]
I'm "translating" from Bun and I noticed that this important optimization is missing.

**Reply by a8m** [reply]
With Ent, you do something like: `client.T.Query().All(ctx)`. In most cases, the count is not known (also by the users) and therefore there's no way to reallocate an array with a fixed size. Also, tbh, this type of allocation (growing a slice) is really minor, especially if you're using ent/gqlgen/sql/etc stack. 

If you really want to do it, you can use the `Select` with `Scan`, but I won't recommend it because it will make your code inconsistent and does not have the same power as `All`.

Another option is to extend the templates with a custom option to determine the size of the array, but again, this minor optimization does not have that much effect compared to all other packages/objects involved in a gqlgen query.

**Reply by frederikhors** [reply]
Ok. And what do you think about an `.Hint()` method for this type of things?

**Reply by frederikhors** [reply]
What worries me when I read "this is a minor optimization" is that this optimization will no longer be dealt with in the future.

Golang in general is one minor optimization after another (for example: https://eng.uber.com/data-race-patterns-in-go/) and I am really put off by the overall performance.

**Reply by frederikhors** [reply]
But luckily there are projects like yours, don't get me wrong.

**Reply by a8m** [reply]
I like when optimizations come out of the box without changing the API. Maybe I'm wrong, but I prefer to keep the API clean even if the cost is a few extra allocations in some code paths.

> What worries me when I read "this is a minor optimization" is that this optimization will no longer be dealt with in the future.

We constantly improve perf in the code (and its generated SQL queries), based on the analysis we do. But prefer to tackle bigger issues before getting to this. In a GraphQL request lifetime on the systems I work on, this is really minor. Again, if it's an issue for you atm, you can use `Select` and `Scan(&entities)`. Or, you can try to use a custom template to expose such API and let us know how it works.

**Reply by frederikhors** [reply]
Yeah. Yeah. I totally agree.
Forgive the little moan (NOT FOR YOUR PROJECT), but for the Go world in general. I'm a little tired.
